internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeserver
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	csicommon "github.com/ceph/ceph-csi/internal/csi-common"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	mount "k8s.io/mount-utils"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultMountPermission = uint64(0o777)
+	// Address of the NFS server.
+	paramServer           = "server"
+	paramShare            = "share"
+	paramClusterID        = "clusterID"
+	mountOptionsField     = "mountOptions"
+	mountPermissionsField = "mountPermissions"
+)
+
+// NodeServer struct of ceph CSI driver with supported methods of CSI
+// node server spec.
+type NodeServer struct {
+	*csicommon.DefaultNodeServer
+	Mounter mount.Interface
+}
+
+// NewNodeServer initialize a node server for ceph CSI driver.
+func NewNodeServer(
+	d *csicommon.CSIDriver,
+	t string,
+	topology map[string]string,
+) *NodeServer {
+	defaultNodeServer := csicommon.NewDefaultNodeServer(d, t, topology)
+
+	return &NodeServer{
+		DefaultNodeServer: defaultNodeServer,
+		Mounter:           defaultNodeServer.Mounter,
+	}
+}
+
+// NodePublishVolume mount the volume.
+func (ns *NodeServer) NodePublishVolume(
```

Comment:
question:- cant we have one NodeStage and multiple NodePublish?

Reply:
> question:- cant we have one NodeStage and multiple NodePublish?

We can but from what i remember from discussion with @nixpanic , 
the kernel remembers nfs mount refs, so if we have one nodePublish for that volume, another nodePublish will just increment the refs and use the same. 
@nixpanic Can you please confirm this.

So there is no particular advantage of having NodeStage and NodePublish. 
and to keep it similar to previously used k8s csi-nfs-driver.

Reply:
Indeed, NFS should be able to detect that the NFS-server and export are already mounted somewhere, and the in-kernel NFS-client is re-used (and refcounted).